utils: add result_try and result_futurize_try

piodul opened the pull request:
Adds `utils::result_try` and `utils::result_futurize_try` - functions which allow to convert existing try..catch blocks into a version which handles C++ exceptions, failed results with exception containers and, depending on the function variant, exceptional futures using the same exception handling logic.

For example, you can convert the following try..catch block:

    try {
        return a_function_that_may_throw();
    } catch (const my_exception& ex) {
        return 123;
    } catch (...) {
        throw;
    }

...to this:

    return utils::result_try([&] {
        return a_function_that_may_throw_or_return_a_failed_result();
    },  utils::result_catch<my_exception>([&] (const Ex&) {
        return 123;
    }), utils::result_catch_dots([&] (auto&& handle) {
        return handle.into_result();
    });

Similarly, `utils::result_futurize_try` can be used to migrate `then_wrapped` or `f.handle_exception()` constructs.

As an example of the usability of the new constructs, two places in the current code which need to simultaneously handle exceptions and failed results are converted to use `result_try` and `result_futurize_try`.

Results of `perf_simple_query --smp 1 --operations-per-shard 1000000 --write`:

```
127041.61 tps ( 67.2 allocs/op,  14.2 tasks/op,   52422 insns/op)
126958.60 tps ( 67.2 allocs/op,  14.2 tasks/op,   52409 insns/op)
127088.37 tps ( 67.2 allocs/op,  14.2 tasks/op,   52411 insns/op)
127560.84 tps ( 67.2 allocs/op,  14.2 tasks/op,   52424 insns/op)
127826.61 tps ( 67.2 allocs/op,  14.2 tasks/op,   52406 insns/op)

126801.02 tps ( 67.2 allocs/op,  14.2 tasks/op,   52420 insns/op)
125371.51 tps ( 67.2 allocs/op,  14.2 tasks/op,   52425 insns/op)
126498.51 tps ( 67.2 allocs/op,  14.2 tasks/op,   52427 insns/op)
126359.41 tps ( 67.2 allocs/op,  14.2 tasks/op,   52423 insns/op)
126298.27 tps ( 67.2 allocs/op,  14.2 tasks/op,   52410 insns/op)
```

The number of tasks and allocations is unchanged. The number of instructions per operations seems similar, it may have increased slightly (by 10-20) but it's hard to tell for sure because of the noisiness of the results.

Tests: unit(dev)

piodul commented:
Temporarily marked as draft because it depends on https://github.com/scylladb/scylla/pull/10014, which waits in `next` at the moment. I set the base to `next`, so if the https://github.com/scylladb/scylla/pull/10014 is dequeued (hopefully not, fingers crossed), additional commits will from that PR will appear here.

However, the PR itself is ready for review, and comments are welcome.

avikivity commented on utils/result_try.hh:
```diff
@@ -0,0 +1,442 @@
+/// if you use it with a future-returning function. See result_futurize_try
+/// for a version which additionally works with exceptional futures.
+template<typename... Handlers>
+auto result_try(auto&& fun, Handlers&&... handlers) {
```
These should most definitely be constrained. Otherwise a compilation error will be a nightmare.

piodul replied:
Constrained `Fun` and `Handlers` with concepts.

avikivity commented on utils/result_try.hh:
```diff
@@ -0,0 +1,442 @@
+static_assert(ExceptionHandle<failed_result_handle<dummy_result<>>, dummy_result<>>);
+
+template<ExceptionContainerResult R>
+struct exception_ptr_handle {
```
I later saw that you did mean a handle.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -2140,35 +2140,39 @@ future<result<>> storage_proxy::mutate_end(future<result<>> mutate_result, utils
         stats.estimated_write.add(lc.latency());
     }
-    try {
+    return utils::result_futurize_try([&] {
```
Does this increase tasks/allocations?

piodul replied:
I think it shouldn't. The `try` body and `catch` handlers here all return ready futures. `utils::result_futurize_try` probably won't allocate tasks because all futures will be ready, but I'll check.

avikivity replied:
I think the futures won't be ready since there's an execution_stage in there. Maybe you removed a continuation here too?

piodul replied:
`mutate_end` is always called in `then_wrapped`, and given a ready future.

avikivity commented:
Looks reasonable. I did not review the internals deeply, but here the API matters much more than the internals, and it looks good.

I'd like a performance report.

piodul commented:
@avikivity I attached results of perf_simple_query to the cover letter. There are 20-30 more instructions per operation, but the number of tasks and allocations is the same.

piodul commented:
> I just realized that I have been running perf_simple_query without the `--write` flag, and without it the benchmark only runs reads but not writes... The logic modified here touches only the write path. I'll re-run and update the cover letter.

Updated. There seem to be 100-150 more instructions now, and tps looks (slightly?) lower.

avikivity commented:
100-150 instructions for a cosmetic change seem high. It's 0.2%, but still it's more than is reasonable. Any idea where it comes from?

piodul commented on utils/result_try.hh:
```diff
@@ -0,0 +1,442 @@
+    auto handle_exception_from_result(const auto& ex, R& original_result, Continuation&& cont)
+            -> typename Converter::template wrapped_type<R> {
+        if constexpr (std::is_base_of_v<Ex, std::remove_cvref_t<decltype(ex)>>) {
+            if constexpr (std::is_invocable_v<Cb, const Ex&, failed_result_handle<R>>) {
```
Not sure. I think you could pass a `const Ex&` to a function `void foo(Ex ex)`. However it would require a copy constructor to be invoked and I'm not sure how it would work if you have a class derived from `std::exception`.

piodul commented on utils/result_try.hh:
```diff
@@ -0,0 +1,442 @@
+};
+
+template<typename Cb>
+struct result_catcher_dots {
```
We would like to avoid converting failed results to exceptions. If you need to rethrow the exception, you should return `into_result` or `into_future` instead.

One thing that I noticed which is missing but will be useful to me when converting the read path is the ability to pass the the exception/failed result to a promise. I'll add it.

Are there some other uses of `std::exception_ptr` you can see but are not handled by this set of methods?

piodul commented on utils/result_try.hh:
```diff
@@ -0,0 +1,442 @@
+// Check that it is a valid handle with some arbitrary ExceptionContainerResult
+static_assert(ExceptionHandle<exception_ptr_handle<dummy_result<>>, dummy_result<>>);
+
+template<typename Ex, typename Cb>
```
I don't think it will work. For example, some handler could explicitly set its return type to `exceptions::coordinator_result<>` and return the failed result via `return handle.into_result()`. If you try to check it with `failed_result_handle<dummy_result<>>` as argument, then you will get a type mismatch on the return statement.

piodul replied:
I think that the best we can do is to constrain everything in `result_try` and `result_futurize_try`, where you know all the necessary types. I plan to write those constraints.

piodul commented on utils/result_try.hh:
```diff
@@ -0,0 +1,442 @@
+#pragma once
```
Done

piodul commented on utils/result_try.hh:
```diff
@@ -0,0 +1,442 @@
+
+    template<typename F, typename... Args>
+    static auto invoke(F&& f, Args&&... args) {
+        return f(std::forward<Args>(args)...);
```
Done, now I'm using `std::invoke` and I'm forwarding `f` to it.

piodul commented on utils/result_try.hh:
```diff
@@ -0,0 +1,442 @@
+    using wrapped_type = T;
+
+    template<typename T>
+    static auto wrap(T&& t) {
```
I went over `auto`s in this file and I removed most of them. The ones I left either couldn't be removed or I felt that writing down the concrete type would be redundant.

piodul commented on utils/result_try.hh:
```diff
@@ -0,0 +1,442 @@
+
+    // Returns the failed result. Can be used only once as this moves out
+    // the failed result. This should be called once during the handler's
+    // lifetime, preferably as `return handler.into_failure()`.
```
Done

piodul commented on utils/result_try.hh:
```diff
@@ -0,0 +1,442 @@
+
+namespace internal {
+
+// Dummy result satisfying ExceptionContainerResult, for static asserts
```
Rewrote the comment, hopefully it doesn't suggest it anymore.

piodul commented:
v1 -> v2: (previous version [link](https://github.com/piodul/scylla/tree/try-catch-v1))

Review fixes:

- Added copyright statement.
- Added constraints for template parameters of `result_try` and `result_futurize_try`.
- `noop_converter::invoke` now forwards the function argument and `std::invoke`s it.
- Reduced the usage of `auto` to make return types and expression types more clear.
- s/into_failure/into_result/ in one of the comments.
- A bit of cleanup related to `ExceptionHandle` and `dummy_handle`.
- Now, `result_futurize_try` will not preempt before invoking a catch handler if the try body didn't preempt. This makes it safer to capture things from stack in catch handlers.
- Introduced `handler.forward_to_promise` for efficient forwarding the exception/failed result to some result<T>, where T may be a different type than the original result.
- Extended tests to check how `result_futurize_try` behave with non-ready futures.
- Added tests for the `handler.forward_to_promise` method.

Further optimization:

- Now, in `result_futurize_try` exceptions thrown directly from the try body are handled by the catch handlers. Previously, it was captured in a futurize_invoke, rethrown and only then handled by catch handlers.
- Tried to minimize moves and pass handlers/results by reference if possible.
- Reorganized the logic of `result_futurize_try` in an attempt to make it more efficient.

----

I cannot say with certainty what caused the 100-150 instruction slowdown in the previous version, but I think the compiler didn't manage as much as I would wish to. The previous version looked similarly to this:

```c++
auto f = seastar::futurize_invoke(fun);
return f.then_wrapped([...handlers = std::move(handlers)] (future_type fun) {
	/* exception/failed results handling logic goes here. */
});
```

In theory, `then_wrapped` should have a code path which calls the lambda in-line (at least in non-debug mode), however I feel that it didn't happen. Changing the above logic to this and marking `internal::result_handle_available_future` with `[[gnu::always_inline]]` helps:

```c++
if (f.available()) {
    return internal::result_handle_available_future(std::move(f), handlers...);
}
return f.then_wrapped([...handlers = std::move(handlers)] (future_type&& f) mutable -> future_type {
    return internal::result_handle_available_future(std::move(f), handlers...);
});
```

Apart from better performance, this has a nice side effect that if the try body doesn't preempt (returns a ready future), then catch handlers will also be called without preemption. In such case, the catch handlers can capture the outer scope by reference (see `mutate_end`). Previously, `then_wrapped` wouldn't guarantee that in debug mode.

The new version seems to have similar performance to pre-PR version, maybe it's 10-20 instructions slower but it's hard to tell because of the noisiness of the results.
